Enable loading stitched tests in partition mode

junjiemao1 opened the pull request:
This series enables the execution of acrn unit tests atop a partition-mode-configured ACRN hypervisor. The test cases can be stitched together and selected by the command line passed via multiboot protocol. It is thus possible to control the execution and collect the results of multiple test cases by talking to the HV shell provided that VM creation/destruction commands are provided.

Summary of changes:

1. Fixes: segment selector reloading, .bss initialization and .comment/.note discarding
2. Tweaks: binary loading address adjustment
3. Enabling: Raw binary generation and test case stitching

Signed-off-by: Junjie Mao <junjie.mao@intel.com>

junjiemao1 committed:
Makefile: generate raw binaries which can be loaded directly

Raw binaries can be used as a multiboot module passed to ACRN hypervisor by the
physical bootloader. This allows the hypervisor to load a unit test without
additional logic for loading ELF files.

v1 -> v2:
* Do not rely on the code starting at exact 4M. Instead prepending zeros at the
  beginning of the image to locate the code right at the position decided at
  link-time. Due to the impreciseness of stext, the starting address of .text
  section is used as the beginning of the image.

Signed-off-by: Junjie Mao <junjie.mao@intel.com>

junjiemao1 committed:
cstart: initialize .bss

The initialization of .bss is available only when the test is loaded by an ELF
loader, which is not the case when it is started directly by a
multiboot-compliant bootloader.

This patch explicitly clears .bss, except the stack which is already in use when
initializing .bss.

Signed-off-by: Junjie Mao <junjie.mao@intel.com>

junjiemao1 committed:
bugfix: cstart64: update selectors after switching GDT on BSP

There is no guarantee on the exact values in the selectors of data segments,
including DS, ES, FS, GS and SS, at the time an OS is invoked, according to
Section 3.2, Multiboot Specification 0.6.96. Not re-initializing the selectors
can lead to segmentation faults the first time a reload of segment base address
is needed.

Signed-off-by: Junjie Mao <junjie.mao@intel.com>

junjiemao1 committed:
intel-iommu: rename to intel_iommu

When stitching multiple tests into a single executable, names of the test files
will be used in function identifiers which cannot contain any dash.

Signed-off-by: Junjie Mao <junjie.mao@intel.com>

junjiemao1 committed:
make: stitch all tests together

Having each test as a separate binary can easily lead to the explosion of file
numbers, considering the number of cases that will be written for thorough
testing. This also leads to difficulties for running multiple tests in a
partition-mode configuration where guest binaries are always loaded by the
physical bootloader.

This patch introduces a 2-phase link approach to allow the tests to be linked in
either one-per-test or all-in-one. The trick behind the scene is that sources of
each test are first compiled and linked before the following object-level
modifications are conducted.

    1. All symbols are converted local, except the function main().
    2. The symbol main() is duplicated to main_<test>() which has global
       linkage.
    3. The symbol main() is converted weak global.

The edited object files are named following the convention
<test>_prelink.o. Each of them can be compiled with the other libraries to
generate an executable ELF, as is done previously, or linked together to
generate one stitched executable that contains all tests indexed by the
generated case_table.c.

Signed-off-by: Junjie Mao <junjie.mao@intel.com>

junjiemao1 commented:
@fyin1 Please help review this series. Thanks!

fyin1-zz commented on guest/x86/flat.lds:
```diff
@@ -1,8 +1,9 @@
 SECTIONS
 {
-    . = 4M + SIZEOF_HEADERS;
+    . = 4M;
```
Can we make the entry to 4M + SIZEOF_HEADERS + 0xC?

fyin1-zz replied:
Other changes look good to me.

junjiemao1 replied:
SIZEOF_HEADERS may change from time to time (e.g. due to the number of sections in the linked executable). This makes the identification of the entry address impossible as we use an objdump'ed raw binary for partition mode.

How about we have two lds, one for single-test executable that should to be loaded as an ELF and the other for the stitched executable that have a fixed entry address and can be loaded as a raw binary?

fyin1-zz replied:
Two lds is complicated. Can we have a header added at the beginning of each raw binary which include the binary entry in fixed offset?

junjiemao1 replied:
Sorry for the late response.

I have restored the original placement of sections. We can rely on the multiboot header which is always 12-byte and right ahead of the entry point.

Also the .comment and .note sections are discarded as it has been met that ignoring these sections in the linker script leads to a very small bss_start (mostly <1K), which leads the test program to crash when it tries to initialize its .bss section.

junjiemao1 committed:
flat.lds: discard .comment and .note

Compilers like gcc may place information like compiler versions, build IDs,
etc. into dedicate sections called .comment or .note. These sections may be
placed even before the entry point if the linker script does not specify how
these sections shall be handled. This can be problematic as it is assumed that
the beginning of the image is a 12-byte multiboot header, followed by the entry
point of the image. It has also been met that ignoring these sections in the
linker script leads to a very small bss_start (mostly <1K), which leads the test
program to crash when it tries to initialize its .bss section.

Discard these sections as these sections contain compiler/linker-specific
information which is mostyl useless at runtime.

Signed-off-by: Junjie Mao <junjie.mao@intel.com>
